### PR TITLE
[CIS-1136] Make it possible to encrypt/decrypt attachments

### PR DESCRIPTION
**Pull request by dmigach**
This change makes it possible to subclass our implementation of StreamCDNClient to implement their own
Also we make attachments mutable so it's possible to decrypt encrypted attachments
A customer gets names and avatarURLs for users from their own backend for security reasons, so they need to be able to set them.
Also this PR disables a flaky test

### CODE REVIEW

**Review by SagarSDagdu — DISMISSED** [review]
LGTM!

File: Sources/StreamChat/Config/ChatClientConfig.swift [hunk 1]
```diff
@@ -91,15 +91,20 @@ public struct ChatClientConfig {
     /// - Parameter apiKey: The API key of the chat app the `ChatClient` connects to.
     ///
     
-    /// Allows to inject a custom API client for uploading attachments, if not specified `StreamCDNClient` is used
-    public var customCDNClient: CDNClient?
+    public enum CDNClientConfig {
```
Comment (evsaev) [inline Sources/StreamChat/Config/ChatClientConfig.swift:94]: Would be great to have some docs here 👍

File: Sources/StreamChat/Config/ChatClientConfig.swift [hunk 2]
```diff
@@ -91,15 +91,20 @@ public struct ChatClientConfig {
...
     /// Returns max possible attachment size in bytes.
     /// The value is taken from custom `maxAttachmentSize` type custom `CDNClient` type.
     /// The default value is 20 MiB.
     public var maxAttachmentSize: Int64 {
-        if let customCDNClient = customCDNClient {
-            return type(of: customCDNClient).maxAttachmentSize
+        if case let .custom(client) = cdnClientConfig {
+            return type(of: client).maxAttachmentSize
         } else {
             return StreamCDNClient.maxAttachmentSize
```
Comment (evsaev) [inline Sources/StreamChat/Config/ChatClientConfig.swift:109]: Shouldn't we take a type from `cdnClientConfig` here? 🤔

File: Sources/StreamChat/Models/Attachments/ChatMessageAudioAttachment.swift [hunk 3]
```diff
@@ -17,13 +17,13 @@ public struct AudioAttachmentPayload: AttachmentPayload {
...
+    public var title: String?
     /// A link to the audio.
-    public internal(set) var audioURL: URL
+    public var audioURL: URL
     /// The audio itself.
-    public let file: AttachmentFile
+    public var file: AttachmentFile
     /// An extra data.
-    let extraData: [String: RawJSON]?
+    var extraData: [String: RawJSON]?
```
Comment (evsaev) [inline Sources/StreamChat/Models/Attachments/ChatMessageAudioAttachment.swift:26]:
```suggestion
    public var extraData: [String: RawJSON]?
```

File: Sources/StreamChat/Models/Attachments/ChatMessageFileAttachment.swift [hunk 4]
```diff
@@ -17,13 +17,13 @@ public struct FileAttachmentPayload: AttachmentPayload {
...
+    public var title: String?
     /// A link to the file.
-    public internal(set) var assetURL: URL
+    public var assetURL: URL
     /// The file itself.
-    public let file: AttachmentFile
+    public var file: AttachmentFile
     /// An extra data.
-    let extraData: [String: RawJSON]?
+    var extraData: [String: RawJSON]?
```
Comment (evsaev) [inline Sources/StreamChat/Models/Attachments/ChatMessageFileAttachment.swift:26]:
```suggestion
    public var extraData: [String: RawJSON]?
```

File: Sources/StreamChat/Models/Attachments/ChatMessageVideoAttachment.swift [hunk 5]
```diff
@@ -17,13 +17,13 @@ public struct VideoAttachmentPayload: AttachmentPayload {
...
+    public var title: String?
     /// A link to the video.
-    public internal(set) var videoURL: URL
+    public var videoURL: URL
     /// The video itself.
-    public let file: AttachmentFile
+    public var file: AttachmentFile
     /// An extra data.
-    let extraData: [String: RawJSON]?
+    var extraData: [String: RawJSON]?
```
Comment (evsaev) [inline Sources/StreamChat/Models/Attachments/ChatMessageVideoAttachment.swift:26]:
```suggestion
    public var extraData: [String: RawJSON]?
```

File: Sources/StreamChat/Config/ChatClientConfig.swift (hunk 2 above)
Comment (evsaev) [reply Sources/StreamChat/Config/ChatClientConfig.swift:109]: would be great to also have a test for it 👍

**Review by evsaev — COMMENTED** [review]
Looks good, thanks! Let's address comments and we are good to 🚢

**Review by evsaev — COMMENTED** [review]
Let's also have a changelog entry for this 🙏

**Comment by dmigach** [comment]
A customer who requested these change is going to check if it helps them, if it does, we're going to finalise it. In the meantime I'll start addressing the issues 👌

File: Sources/StreamChat/Workers/ChannelUpdater_Tests.swift [hunk 6]
```diff
@@ -425,7 +425,8 @@ class ChannelUpdater_Tests: StressTestCase {
         XCTAssertEqual(apiClient.request_endpoint, AnyEndpoint(referenceEndpoint))
     }
 
-    func test_hideChannel_successfulResponse_isPropagatedToCompletion() throws {
+    // TODO: Fix and enable
+    func _test_hideChannel_successfulResponse_isPropagatedToCompletion() throws {
```
Comment (nuno-vieira) [inline Sources/StreamChat/Workers/ChannelUpdater_Tests.swift:429]: Please remove this, it seems I was able to at least make it more stable on https://github.com/GetStream/stream-chat-swift/pull/1314

**Comment by nuno-vieira** [comment]
> A customer who requested these change is going to check if it helps them, if it does, we're going to finalise it. In the meantime I'll start addressing the issues 👌

Any news about this one?

**Comment by kojiba** [comment]
This decrease code readability cause when reader sees `var` he assumes its changing somewhere. 
Maybe there is options how to do this extendable but in more smart way, instead of just making everything public and var?

**Comment by kojiba** [comment]
For example we could introduce a new delegate for ChannelController with will have a method 
`process(attachment: AttachmentInputType) -> AttachmentOutputType`
for input we can give a struct, and for output we could require just some proto to display data

File: Sources/StreamChat/Config/ChatClientConfig.swift (hunk 1 above)
Comment (dmigach) [reply Sources/StreamChat/Config/ChatClientConfig.swift:94]: It seems like we might want to drop this change in the end if the customer will be able to achieve e2e without it

File: Sources/StreamChat/Config/ChatClientConfig.swift [hunk 7]
```diff
@@ -171,3 +176,22 @@ public struct APIKey: Equatable {
...
+            return client
+        }
+        set {
+            if let customClient = newValue {
+                cdnClientConfig = .custom(client: customClient)
+            } else {
+                cdnClientConfig = .default(type: StreamCDNClient.self)
+            }
+        }
+    }
```
Comment (nuno-vieira) [inline Sources/StreamChat/Config/ChatClientConfig.swift:196]: No need to add deprecations, V4 will include breaking changes, so we can break the API for now 👍 Anyways, all deprecations should live in the `Deprecations.swift` file.

**Review by nuno-vieira — CHANGES_REQUESTED** [review]

**Comment by dmigach** [comment]
> > A customer who requested these change is going to check if it helps them, if it does, we're going to finalise it. In the meantime I'll start addressing the issues 👌
> 
> Any news about this one?

The customer was happy with the suggested solution, and they won't need to deal with CDN in the end. They went for implementing it so we might expect additional request in the nearest future

**Review by tbarbugli — APPROVED** [review]

**Review by tbarbugli — APPROVED** [review]

**Pull request merged**